### PR TITLE
Add update statement support

### PR DESCRIPTION
**Pull request by gruuya**
Implementation of `UPDATE` statement in Seafowl, along similar lines as `DELETE`.

Basically, we pinpoint the exact partitions that need to be updated, and then for the new table version we do a projection of all columns that were changed using the provided assignment expressions. If a WHERE qualifier has been used, we further wrap up the assignment expressions using a CASE expression, to scope down the changes only to affected rows. The columns which were not changed are just projected as is.

### CODE REVIEW

File: src/context.rs [hunk 1]
```diff
@@ -974,25 +997,31 @@ impl SeafowlContext for DefaultSeafowlContext {
...
                         None => None,
                         Some(expr) => Some(query_planner.sql_to_rex(expr, &table_schema, &mut HashMap::new())?),
                     };
 
-                    let assignments_expr = assignments.iter().map(|a| {
-                        Ok(Assignment { column: compound_identifier_to_column(&a.id)?, expr: query_planner.sql_to_rex(a.value.clone(), &table_schema, &mut HashMap::new())? })
-                    }).collect::<Result<Vec<Assignment>>>()?;
+                    let assignment_exprs = assignments.iter().map(|a| {
+                        Ok((
+                            table_schema.field_with_unqualified_name(&normalize_ident(&a.id[0]))?.name().clone(),
```
Comment (gruuya) [inline src/context.rs:1015]: I opted for interpreting the idents as single column names, and not as compund identifiers, as it made more sense for an `UPDATE` statement.

File: src/context.rs [hunk 2]
```diff
@@ -1237,52 +1266,229 @@ impl SeafowlContext for DefaultSeafowlContext {
...
+                                    }
+                                })
+                                .collect::<Result<Vec<(Arc<dyn PhysicalExpr>, String)>>>(
+                                )?;
+
+                            let mut final_partition_ids =
+                                Vec::with_capacity(partitions.len());
+
+                            // Group adjacent partitions eligible for updating
+                            let partitions_grouped: Vec<(bool, Vec<SeafowlPartition>)> =
```
Comment (gruuya) [inline src/context.rs:1383]: In principle, if the ordering is not important, then we can (here as well as for DELETE) divide all the partitions into exactly two groups, and re-process the group that needs to be changed in one go.

File: src/context.rs (hunk 1 above)
Comment (mildbyte) [reply src/context.rs:1015]: That makes sense, compound identifiers are only useful in an `UPDATE ... USING` (where you might update with some values derived from a different table)

File: tests/end_to_end.rs [hunk 3]
```diff
@@ -1316,3 +1329,117 @@ async fn test_delete_statement() {
...
+        "|                 |                |                  |                     | 42         |",
+        "|                 | 5555           |                  | 2022-01-01 20:01:01 | 42         |",
+        "|                 | 5555           |                  | 2022-01-01 20:02:02 | 42         |",
+        "|                 | 3333           |                  | 2022-01-01 20:03:03 | 42         |",
+        "|                 | 5555           |                  |                     | 42         |",
+        "|                 | 5555           |                  |                     | 42         |",
+        "|                 |                |                  |                     | 42         |",
+        "+-----------------+----------------+------------------+---------------------+------------+",
+    ];
+    assert_batches_eq!(expected, &results);
```
Comment (mildbyte) [inline tests/end_to_end.rs:1444]: These tests look good. Worth testing some extra error/corner cases here:

  - requires an implicit cast (e.g. `SET some_time = '2022-01-01 21:21:21')
  - results in a type mismatch (e.g. `SET some_int_value = 'nope'`)
  - references a nonexistent column (`SET nonexistent = 42`)
  - doesn't change anything (`SET some_bool_value = TRUE WHERE some_value = 200`)
  - redundant update (`SET some_bool_value = TRUE, some_bool_value = (some_int_value = 5555)`)
  - a more complex expression (`SET some_other_value = CASE WHEN some_int_value = 5555 THEN 'five' WHEN some_int_value = 3333 THEN 'three' ELSE 'none' END`)

File: src/context.rs [hunk 4]
```diff
@@ -1237,52 +1266,229 @@ impl SeafowlContext for DefaultSeafowlContext {
...
+                                        update_plan = Arc::new(ProjectionExec::try_new(
+                                            projection_exprs.clone(),
+                                            scan_plan,
+                                        )?);
+
+                                        println!("table schema: {:?}", table.schema());
+                                        println!(
+                                            "update schema: {:?}",
+                                            update_plan.schema()
+                                        );
```
Comment (mildbyte) [inline src/context.rs:1470]:
```suggestion
                                        debug!("table schema: {:?}", table.schema());
                                        debug!(
                                            "update schema: {:?}",
                                            update_plan.schema()
                                        );
```

File: src/context.rs (hunk 2 above)
Comment (mildbyte) [reply src/context.rs:1383]: Kinda true (especially since we're reordering data inside of the partitions anyway, since we're merging changed with the unchanged rows), but we have this code already, so I don't mind it staying here.

**Review by mildbyte — COMMENTED** [review]
Looking good! I added suggestions for some more tests that are worth writing to expose potential issues.

File: src/context.rs (hunk 4 above)
Comment (gruuya) [reply src/context.rs:1470]: Thanks, I'll rather remove this completely, this was just for debugging.

File: tests/end_to_end.rs (hunk 3 above)
Comment (gruuya) [reply tests/end_to_end.rs:1502]: Good idea, added all these variations now.

I've commented out the type mismatch tests for now because, even though they themselves pass, the last case errors out during planning with `Internal("Internal SQL error: \"error returned from database: syntax error at end of input\"")`. Looks like there's something not being closed/cleaned from our side while planing the type mismatch queries.

Comment (gruuya) [reply tests/end_to_end.rs:1500]: Turns out the issue was that when planning UPDATE we first create a new table version, and only then try to execute the plan into partitions, which fails due to the type mismatch, but leaves the new empty table version hanging.

I've fixed it by creating the new table version only when needed (when linking to newly created partitions).

File: src/utils.rs [hunk 5]
```diff
@@ -90,6 +92,22 @@ pub async fn gc_partitions(context: &DefaultSeafowlContext) {
     }
 }
 
+// Group adjacent partitions using the provided function
+pub fn group_partitions<F>(
+    partitions: Vec<SeafowlPartition>,
+    f: F,
+) -> Vec<(bool, Vec<SeafowlPartition>)>
+where
+    F: FnMut(&SeafowlPartition) -> bool,
```
Comment (mildbyte) [inline src/utils.rs:101]: Can this be `Fn` instead of `FnMut`, given that it doesn't mutate `partitions_to_filter`/`partitions_to_update`? Not sure if this will result in any compiler optimizations though.

File: tests/end_to_end.rs [hunk 6]
```diff
@@ -1316,3 +1329,173 @@ async fn test_delete_statement() {
...
+
+    // This one's a bit different
+    let err = context
+        .plan_query("UPDATE test_table SET some_other_value = 'nope'")
+        .await
+        .unwrap_err();
+
+    assert!(err
+        .to_string()
+        .contains("Unsupported CAST from Utf8 to Decimal128(38, 10)"));
```
Comment (mildbyte) [inline tests/end_to_end.rs:1458]: Is it possible to break these out into a separate test (e.g. `test_update_errors`) just for the sake of higher test precision?

(though OTOH we do test that we can perform an update after an error, so it's worth keeping one error case here)

File: src/context.rs [hunk 7]
```diff
@@ -1237,52 +1265,170 @@ impl SeafowlContext for DefaultSeafowlContext {
...
+                                    } else {
+                                        // Just replicate the omitted columns
+                                        Ok((
+                                            col(f.name(), &schema)?,
+                                            f.name().to_string(),
+                                        ))
+                                    }
+                                })
+                                .collect::<Result<Vec<(Arc<dyn PhysicalExpr>, String)>>>(
+                                )?;
```
Comment (mildbyte) [inline src/context.rs:1370]: I think this `assignments -> projection exprs` function is self-contained enough that it could be factored out (and potentially unit tested) to keep this UPDATE handler shorter

**Review by mildbyte — APPROVED** [review]

File: src/utils.rs (hunk 5 above)
Comment (gruuya) [reply src/utils.rs:101]: Yup, changing.

File: src/context.rs (hunk 7 above)
Comment (gruuya) [reply src/context.rs]: Good point, extracted, though note that unti testing is a bit of a pain due to need to compare trait objects for equality.
